Documentation/Evolution/CharacterClasses.md

```diff
@@ -0,0 +1,354 @@
+
+```swift
+let pattern = /^Caf.$/.unicodeScalarSemantics
+str.contains(pattern)   // false
+```
+
+Initially, we plan to allow configuration between character, Unicode scalar, and POSIX-compatible character class matching.
+
+
+### Comprehensibility
+
+For each of the character classes described below, we will base the  also add corresponding Boolean properties on both `Character` and `UnicodeScalar`. These properties will provide multiple valuable benefits:
+
+- they will be accessible from within the regular expression/pattern DSL,
+- the characters or values that each class matches can be clearly described in the symbols' documentation, and
+- the matched characters can be validated by accessing the properties directly.
+
+For example, the `\w` character class, for "word" characters, will have a matching `matchesAsWordCharacter` property.
+
+```swift
+"C".matchesAsWordCharacter  // true
+"é".matchesAsWordCharacter  // true
+"‡".matchesAsWordCharacter  // false
+```
+
+We could further extend the API to allow the redefinition of individual character classes, such as making `\d` match only the characters in the range `0` through `9`.
+
+```swift
+let fiveDigits = /\d/.customizingClass(\.matchesAsDigit, toMatch: /[0-9]/)
+```
+
+## Detailed Design
+
+The following sections detail the different character classes that will be supported by Swift regular expression literals, along with their proposed counterparts in the regular expression DSL.
+
+### Literal characters
+
+A literal character (such as `a`, `é`, or `한`) in a regex literal matches that particular character or code sequence. When matching in Unicode scalar or POSIX mode, the underlying code sequence must be an exact match.
+
+```swift
+let str = "Cafe\u{301}"
+str.contains(/e/)                           // false
+str.contains(/e/.unicodeScalarSemantics)    //  true
+```
+
+**DSL:** Literal characters are represented by string literals.
+
+```swift
+Pattern {
+    "abc"
+}
+```
+
+### Match any: `.`, `\X`
+
+The dot metacharacter matches any single character or element except for a newline (see the _Whitespace and Newlines_ section, below). In "dot matches all" mode, newlines are treated like any other character. A future proposal will discuss modes in more detail.
+
+`\X` matches any grapheme cluster (`Character`), even when the regular expression is otherwise matching at the Unicode scalar or POSIX level.
+
+**DSL:** The "match anything" pattern is spelled `.anyCharacter`. `.anyCharacter` will match at the level of the overall pattern unless given a more specific level, so `.anyCharacter.graphemeScalarSemantics` is equivalent to the `\X` metacharacter.
+
+```swift
+Pattern {
+    OneOrMore(.anyCharacter)
+}
+```
+
+### Digits: `\d`,`\D`
+
+`\d` matches a "digit", which is any `Character` or Unicode scalar with the Unicode general category `Decimal_Number`.
```

Comment:
What's the difference between Character.isNumber and a digit? 

I think all of these would make sense to propose the Character and Unicode.Scalar API first, then show how it corresponds to a built-in character classes in regex and an example of what it might look like in `Pattern`, noting that `Pattern` is still under development.